packages/eslint-plugin/tests/rules/no-unsafe-declaration-merging.test.ts

```diff
@@ -0,0 +1,71 @@
+import rule from '../../src/rules/no-unsafe-declaration-merging';
+import { getFixturesRootDir, RuleTester } from '../RuleTester';
+
+const rootPath = getFixturesRootDir();
+
+const ruleTester = new RuleTester({
+  parser: '@typescript-eslint/parser',
+  parserOptions: {
+    sourceType: 'module',
+    tsconfigRootDir: rootPath,
+    project: './tsconfig.json',
+  },
+});
+
+ruleTester.run('no-unsafe-declaration-merging', rule, {
+  valid: [
+    `
+interface Foo {}
+class Bar implements Foo {}
+    `,
+    `
+namespace Foo {}
+namespace Foo {}
+    `,
+    `
+enum Foo {}
+namespace Foo {}
+    `,
+    `
+namespace Fooo {}
+function Foo() {}
+    `,
+    `
+const Foo = class {};
+    `,
+  ],
+  invalid: [
+    {
+      code: `
+interface Foo {}
+class Foo {}
+      `,
+      errors: [
+        {
+          messageId: 'unsafeMerging',
+        },
+        {
+          messageId: 'unsafeMerging',
+        },
+      ],
+    },
+    {
+      code: `
+namespace Foo {
+  export interface Bar {}
+}
+namespace Foo {
+  export class Bar {}
+}
+      `,
+      errors: [
+        {
+          messageId: 'unsafeMerging',
+        },
+        {
+          messageId: 'unsafeMerging',
+        },
+      ],
+    },
+  ],
+});
```

Comment:
These tests are quite good for the main cases, but don't cover very many edge cases. How about some that cover...

* Declaring a class within odd places:
    * Inside a function declaration and IIFE
    * Inside a class extends clause, like `class Apple extends class Banana {}`
* Inside a `declare global`